### ISSUE TITLE
Maximum integer part and scale

### ISSUE THREAD

**Issue by nesquikm** [onepub-dev/money.dart#75]
I couldn't find any mention of any restrictions, but as far as I can see the integer and fractional parts are represented by the BigInt. However, the simple test below breaks when maxScale>18 or maxInts>18:

```dart
import 'package:test/test.dart';
import 'package:money2/money2.dart';

void main() {
  const maxScale = 18;
  const maxInts = 18;

  setUp(() {
    for (var scale = 0; scale <= maxScale; scale++) {
      final c = Currency.create('C$scale', scale, symbol: '=$scale=');
      Currencies().register(c);
    }
  });

  test('scale 0-$maxScale test', () {
    for (var scale = 0; scale <= maxScale; scale++) {
      final c = Currencies().find('C$scale');
      expect(c, isNotNull);
      final str = scale == 0 ? '1' : '1.${'0' * (scale - 1)}1';
      final fmt = scale == 0 ? 'S#' : 'S#.${'#' * scale}';
      expect(Money.parseWithCurrency(str, c!).format(fmt), '=$scale=$str', reason: 'Failed with $scale scale');
    }
  });

  test('integers 0-$maxInts test', () {
    for (var ints = 0; ints <= maxInts; ints++) {
      final c = Currencies().find('C0');
      expect(c, isNotNull);
      final str = ints == 0 ? '0' : '9' * ints;
      final fmt = 'S#';
      expect(Money.parseWithCurrency(str, c!).format(fmt), '=0=$str', reason: 'Failed with $ints ints');
    }
  });

  test('scale 0-$maxScale and integers 0-$maxInts test', () {
    for (var scale = 0; scale <= maxScale; scale++) {
      for (var ints = 0; ints <= maxInts; ints++) {
        final c = Currencies().find('C$scale');
        expect(c, isNotNull);
        final intsStr = ints == 0 ? '0' : '9' * ints;
        final str = scale == 0 ? intsStr : '$intsStr.${'0' * (scale - 1)}1';
        final fmt = scale == 0 ? 'S#' : 'S#.${'#' * scale}';
        expect(Money.parseWithCurrency(str, c!).format(fmt), '=$scale=$str',
            reason: 'Failed with $scale scale, $ints ints');
      }
    }
  });
}

```

**Comment by bsutton** [comment]
So I'm uncertain what action your are looking to be taken?

**Comment by nesquikm** [comment]
There is no mention of border conditions in the documentation except [AmountTooLargeException](https://pub.dev/documentation/fixed/latest/fixed/AmountTooLargeException-class.html) which is confusing because the library does not work correctly even at smaller _scale_ and without _num_. First of all, I was looking for an answer to the question at what point does the library stopping working in my test, assuming that I missed something.

**Comment by bsutton** [comment]
@nesquikm so I ran your unit tests and they all pass?

**Comment by bsutton** [comment]
What is your test environment?
I'm running 64bit linux - ubuntu 22.04
dart 2.19.5

**Comment by nesquikm** [comment]
As I mention before, I offered Fixed.parse() as a fix for Money.parseWithCurrency() bug. You just added test for [Fixed.parse()](https://github.com/onepub-dev/money.dart/commit/68e8c5dd64d740f0b476c8c9c592932f2247ec2c), so it's should pass :)

**Comment by bsutton** [comment]
sorry I missed that.

So we need the same set of unit tests for Money?

**Comment by nesquikm** [comment]
Yeah, there is [test for money](https://github.com/broxus/money2_improver/blob/main/test/money2_test.dart) (also you can [find it in the first comment](https://github.com/onepub-dev/money.dart/issues/75#issue-1613146722)), try to increase maxScale or maxInts to see the problem.

And [this is my temporary fix](https://github.com/broxus/money2_improver/blob/main/lib/src/money2_improver_ext.dart).

**Comment by bsutton** [comment]
do you have a specific requirement?
Otherwise, I'm inclined to just document the limit.

**Comment by bsutton** [comment]
I've documented the limits here:
https://money2.onepub.dev/

and I've opened an enhancement request to increase the limits.

**Comment by nesquikm** [comment]
I just found that Money.fromJson and toJson have similar problems (they use toInt()). So I added some tests and updated my package that fixes all of these problems: [https://pub.dev/packages/money2_fixer](https://pub.dev/packages/money2_fixer)

For context you can run [example](https://github.com/nesquikm/money2_fixer/blob/main/example/money2_improver_example.dart):

```text
Test with: integers: 3, decimals: 18
String: 999.999999999999999999
Pattern: 0.################## S
Parse and format:
  with Money.parse: 999.999999999999999999 SV
  with improved method: 999.999999999999999999 SV
To and from json:
 parsing, integer part: 999, decimal part: 999999999999999999
  to and from Money.json: integer part: 999  decimal part: 999999999999999999
  to and from json improved: integer part: 999  decimal part: 999999999999999999

Test with: integers: 3, decimals: 19
String: 999.9999999999999999999
Pattern: 0.################### S
Parse and format:
  with Money.parse: -842.8297329635842064385 SV
  with improved method: 999.9999999999999999999 SV
To and from json:
 parsing, integer part: 999, decimal part: 9999999999999999999
  to and from Money.json: integer part: 999  decimal part: 9223372036854775807
  to and from json improved: integer part: 999  decimal part: 9999999999999999999

Test with: integers: 19, decimals: 3
String: 9999999999999999999.999
Pattern: 0.### S
Parse and format:
  with Money.parse: 9223372036854775807.999 SV
  with improved method: 9999999999999999999.999 SV
To and from json:
 parsing, integer part: 9999999999999999999, decimal part: 999
  to and from Money.json: integer part: 9223372036854775807  decimal part: 999
  to and from json improved: integer part: 9999999999999999999  decimal part: 999
```

**Comment by bsutton** [comment]
So instead of doing that raise a pr.

Unnecessary bifurcation of dary packages is not helping the dart community.

**Comment by nesquikm** [comment]
You are of course right, I will try to make a PR as soon as I have time.
According to statistics, many users are already using money2_fixer, so I need to fix it as soon as possible.

**Comment by bsutton** [comment]
can I suggest that we merge the two projects and I add you as a contributor to money2 and then mark money2_fixer as deprecated.

I'm always happy to have help.

**Comment by nesquikm** [comment]
Yes, of course. It would probably be nice to add tests from money2_fixer too.

**Comment by bsutton** [comment]
That would be great!

**Comment by nesquikm** [comment]
So, I created two PR:

https://github.com/onepub-dev/fixed/pull/22
https://github.com/onepub-dev/money.dart/pull/91

But I think we should eliminate all the rest toInt()/fromInt(), and check math operations for big int support.
